### PR TITLE
Parallelise chunk processing

### LINKED ISSUE #121: Parallelise requests to S3 active storage
Currently when using S3 storage backend, requests to the S3 active storage server are made sequentially for each chunk, waiting for the result of one request until the next is sent. This is functional, but not performant. Requests should be made concurrently, avoiding multiple round trip latencies to the active storage server and from there to the S3 object store. It also allows us to exploit parallelism in the active storage server.

This could be implemented in a number of ways:

* [multiple processes](https://docs.python.org/3/library/multiprocessing.html)
* [multiple threads](https://docs.python.org/3/library/threading.html)
* [asyncio](https://docs.python.org/3/library/asyncio.html)

Of these, asyncio is likely to be the most performant, but requires the library to operate with an async event loop. This may or may not be desirable for the user of the library. Threading typically has a lower overhead than multiprocessing, and may be a good compromise.

We may need to consider throttling requests to the active storage server, particularly until memory management features are implemented in the server. Excessive throttling could have a negative impact on performance however, so this should be used with caution.

### PR DESCRIPTION
**Pull request by markgoddard**
Previously, each storage chunk was processed sequentially. This is
functional, but not performant, particularly when using Reductionist
(S3), due to the sequential round trip latencies of requests. This means
we are not taking advantage of the concurrency provided by Reductionist.

This change switches to use a threadpool executor to process chunks.
This is unlikely to be as performant as asyncio, but still provides a
significant improvement without requiring an async runtime.

Some basic throttling of requests is provided by using a threadpool of a
configurable size. This works well when using a single Active object,
but may need refinement if multiple Active objects are in use at the
same time to avoid overloading the Reductionist server.

In order to benefit from connection reuse, we switch to using a requests
Session object that is shared by all threads in the pool.

This change also adds support for specifying a CA certificate to use
when making requests to the Reductionist API.

Closes #121

### CODE REVIEW

**Review by valeriupredoi — APPROVED** [review]
@markgoddard this looks great, very many thanks! :beer: A couple points:
- I am approving but can we delay the merge by just a few days, just had a call with @kchasapis and the guys at IME/RED are starting to understand the mechanics and use of `Active` - Konstantinos is now forking the rpo and would probably appreciate it if there were no new arguments and functionality of Active that he'd have to make use of on RED; so I'll merge next week, should be good by then
- would it make sense to start thinking of a configuration that contains, amongst others, `max_threads`?
- I've used the `minio_start` call we have in `main` rather than yours in here (the one in `main` is still yours haha) - lemme know if the one here was better, so I can revert to it :+1:

File: activestorage/reductionist.py [hunk 1]
```diff
@@ -144,12 +157,11 @@ def build_request_data(source: str, bucket: str, object: str, offset: int,
...
 
 
-def request(url: str, username: str, password: str, request_data: dict):
+def request(session: requests.Session, url: str, request_data: dict):
     """Make a request to a Reductionist API."""
-    response = requests.post(
+    response = session.post(
         url,
         json=request_data,
-        auth=(username, password)
```
Comment (valeriupredoi) [inline activestorage/reductionist.py:152]: missed this one: will this be a problem when we get out of Minio and start using the actual AWS S3?

Comment (markgoddard) [reply activestorage/reductionist.py:152]: No functional change here - the auth parameters are moved from individual requests to the session.

**Comment by markgoddard** [comment]
> @markgoddard this looks great, very many thanks! 🍺 A couple points:
> 
> * I am approving but can we delay the merge by just a few days, just had a call with @kchasapis and the guys at IME/RED are starting to understand the mechanics and use of `Active` - Konstantinos is now forking the repo and would probably appreciate it if there were no new arguments and functionality of Active that he'd have to make use of on RED; so I'll merge next week, should be good by then

No problem for me to wait here. I can always use the branch if necessary.

> * would it make sense to start thinking of a configuration that contains, amongst others, `max_threads`?

The configuration of PyActiveStorage does need improvement. The current config.py method isn't really a good long-term solution since it's part of the Python module. This probably isn't the right place to discuss it though.

> * I've used the `minio_start` call we have in `main` rather than yours in here (the one in `main` is still yours haha) - lemme know if the one here was better, so I can revert to it 👍

No problem - whatever works.

File: activestorage/reductionist.py (hunk 1 above)
Comment (valeriupredoi) [reply activestorage/reductionist.py:152]: got it, thanks! :+1:

**Comment by valeriupredoi** [comment]
awesome, and yes, configuration needs discussion (even more so after RED integration), not here. Not a worry, won't dillydally anymore, will merge on Monday :+1:

**Comment by valeriupredoi** [comment]
cheers @markgoddard :beer:

**Pull request merged**